lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -602,43 +603,23 @@ func (c *Client) handleBitmap(cb *C.CGOBitmap) C.CGOErrCode {
 	uptr := (*uint8)(ptr)
 	data := unsafe.Slice(uptr, int(cb.data_len))
 
-	// Convert BGRA to RGBA. It's likely due to Windows using uint32 values for
-	// pixels (ARGB) and encoding them as big endian. The image.RGBA type uses
-	// a byte slice with 4-byte segments representing pixels (RGBA).
-	//
-	// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/8ab64b94-59cb-43f4-97ca-79613838e0bd
-	//
-	// Also, always force Alpha value to 100% (opaque). On some Windows
-	// versions (e.g. Windows 10) it's sent as 0% after decompression for some reason.
-	for i := 0; i < len(data); i += 4 {
-		data[i], data[i+2], data[i+3] = data[i+2], data[i], 255
-	}
-
-	rect := image.Rectangle{
-		Min: image.Pt(int(cb.dest_left), int(cb.dest_top)),
-		Max: image.Pt(int(cb.dest_right)+1, int(cb.dest_bottom)+1),
-	}
-
-	var img *image.RGBA
-	isFullBitmap := cb.dest_right-cb.dest_left == 63 &&
-		cb.dest_bottom-cb.dest_top == 63
-	if isFullBitmap {
-		if c.img == nil {
-			c.img = image.NewRGBA(rect)
-		} else {
-			c.img.Rect = rect
-		}
-		img = c.img
-	} else {
-		img = image.NewRGBA(rect)
-	}
-
-	copy(img.Pix, data)
+	buf := make([]byte, 0, 21)
+	buf = append(buf, byte(tdp.TypePNG2Frame))
+	buf = binary.BigEndian.AppendUint32(buf, uint32(len(data)))
+	buf = binary.BigEndian.AppendUint32(buf, uint32(cb.dest_left))
+	buf = binary.BigEndian.AppendUint32(buf, uint32(cb.dest_top))
+	buf = binary.BigEndian.AppendUint32(buf, uint32(cb.dest_right))
+	buf = binary.BigEndian.AppendUint32(buf, uint32(cb.dest_bottom))
 
-	if err := c.cfg.Conn.WriteMessage(tdp.NewPNG(img, c.cfg.Encoder)); err != nil {
-		c.cfg.Log.Errorf("failed to send PNG frame %v: %v", img.Rect, err)
+	if _, err := c.cfg.Conn.Write(buf); err != nil {
+		c.cfg.Log.Error("failed to write PNG header")
 		return C.ErrCodeFailure
 	}
+
+	if _, err := c.cfg.Conn.Write(data); err != nil {
```

Comment:
Writing the header separate from the PNG data might save you a tiny bit because you don't have to concatenate them, but either way we still have 1 copy. 

It used to be a copy into `img.Pix`, now it's a copy into into a single `[]byte`.

You can also eliminate the per-bitmap allocation by reusing the same buffer.

To summarize, I would:

1. Add a `bitmapBuffer []byte` to `Client`

2. When receiving a PNG from Rust, append the left/right/bottom/top directly to this rather than allocating a one-time use `buf`:

```
c.bitmapBuf = c.bitmapBuf[:0]
c.bitmapBuf = binary.BigEndian.AppendUint32(c.bitmapBuf, ...)
c.bitmapBuf = binary.BigEndian.AppendUint32(c.bitmapBuf, ...)
```

3.  Perform a copy from the Rust-owned `data` into this bitmap buffer:
```
c.bitmapBuf = append(c.bitmapBuf, data...)
```

4. Now you can change the internal structure of `PNG2Frame` so that `Encode` does nothing.
```
type PNG2Frame {
    msg []byte // already encoded
}

func (f PNG2Frame) Encode() ([]byte, error) {
    return f.msg, nil
}
```

5. Now you can go back to using WriteMessage, which will capture the recording data.

Reply:
Thanks for the info! I had similar idea in mind, but I was struggling to find a way to how to effectively move data without much coping & allocations in Go.